src/main/host/thread_ptrace.c

```diff
@@ -297,69 +299,41 @@ static const char* _syscall_regs_to_str(const struct user_regs_struct* regs) {
+    }
+    StopReason reason = _getStopReason(wstatus);
+    if (reason.type != STOPREASON_SIGNAL) {
+        error("Unexpected stop reason: %d", reason.type);
+    }
+    if (reason.signal.signal != SIGTRAP) {
+        error("Unexpected signal: %d", reason.signal.signal);
+    }
 
-#ifdef SHADOW_COVERAGE
-    // The instrumentation in coverage mode causes corruption in between vfork
-    // and exec. Use fork instead.
-    pid_t pid = fork();
-#else
-    pid_t pid = vfork();
-#endif
+    // Stop and detach the child, allowing the shadow worker thread to
+    // attach it when it's run.
+    if (ptrace(PTRACE_DETACH, child, 0, SIGSTOP) < 0) {
+        error("ptrace: %s", g_strerror(errno));
+    }
+}
 
-    switch (pid) {
-        case -1: {
-            error("fork: %s", g_strerror(errno));
-            return -1;
-        }
-        case 0: {
-            // child
-
-            // CAUTION: Because we used `vfork`, we still share memory of the
-            // parent process (which will be suspended until we call `exec`).
-            // i.e. be wary of any potential side-effects to global variables
-            // etc.
- 
-            // Ensure that the child process exits when Shadow does.  Shadow
-            // ought to have already tried to terminate the child via SIGTERM
-            // before shutting down (though see
-            // https://github.com/shadow/shadow/issues/903), so now we jump all
-            // the way to SIGKILL.
-            if (prctl(PR_SET_PDEATHSIG, SIGKILL) < 0) {
-                error("prctl: %s", g_strerror(errno));
-                exit(1);
-            }
-            // Validate that Shadow is still alive (didn't die in between forking and calling
-            // prctl).
-            if (getppid() != shadow_pid) {
-                error("parent (shadow) exited");
-                exit(1);
-            }
-            // Disable RDTSC
-            if (prctl(PR_SET_TSC, PR_TSC_SIGSEGV, 0, 0, 0) < 0) {
-                error("prctl: %s", g_strerror(errno));
-                return -1;
-            }
-            // Become a tracee of the parent process.
-            if (ptrace(PTRACE_TRACEME, 0, 0, 0) < 0) {
-                error("ptrace: %s", g_strerror(errno));
-                return -1;
-            }
-            // Because we're now being ptraced, execvpe will put this process
-            // in a ptrace-stop.  Luckily, exec will re-awakens the parent
-            // before stopping this one.
-            if (execvpe(file, argv, envp) < 0) {
-                error("execvpe: %s", g_strerror(errno));
-                return -1;
-            }
-        }
-        default: {
-            // parent
-            info("started process %s with PID %d", file, pid);
-            return pid;
-        }
+// To be called in plugin child processes before `exec`
+void _childPreExecCb() {
+    // Disable RDTSC
+    if (prctl(PR_SET_TSC, PR_TSC_SIGSEGV, 0, 0, 0) < 0) {
+        error("prctl: %s", g_strerror(errno));
```

Comment:
Does anything bad happen by accessing the logger from within the child of vfork? And if the child process aborts due to `error()`, does the parent process also abort?